design-proposals/external-monitoring.md

```diff
@@ -0,0 +1,173 @@
+This effort is very important as a good monitoring solution & design directly impacts an operator's quality.
+For example, operator framework define the [Operator Capability Levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/),
+as a metric of an operator's maturity. Looking at the current state of Kubevirt, arguably we fill the requirements
+of the first 3 levels, which are "Basic install", "Seamless Upgrades" and "Full Lifecycle" management. That said,
+we probably have some more work to do in order to become level 4 and 5, which are "Deep insights" and "Auto Pilot".
+These two levels are tightly related to our ability to perform monitoring. When we would fulfil level 4 with deep
+insights, we would carve the way to being able to auto-config/heal/scale the cluster, therefore becoming an "auto-pilot"
+operator.
+
+**TLDR**: I suggest a monitoring approach in which the monitoring entities are **autonomous and external** in terms
+of development (a dedicated repository) and deployment (a dedicated controller / operator).
+
+**This design proposal aims to be a high-level design for a general path forward.** In other words, during this design
+discussions, I prefer not to dwell on implementation details. Instead, I would like this PR to carve the way for an
+agreed upon, very high-level, long-term design. This design will serve as a long-term vision and path forward for our monitoring
+efforts. Follow-up design proposals should be created in order to deep-dive into the implementation of the components that
+would be agreed upon in this PR.
+
+## Motivation
+
+### Development speed, and the ability to propose new ideas quickly
+
+Monitoring structure aims to be similar across different repositories. Therefore,
+there is a major overhead of aligning with different components. For example, let's say that we decide
+that similar alerts need to pop up from all of our Kubevirt repos, let's say that it's a `component_X_is_unhealthy`.
+In other words, we would have `component_virt_handler_is_unhealthy`, `component_virt_controller_is_unhealthy` etc for
+kubevirt/kubevirt, and similar alert for kubevirt/CDI, kubevirt/HCO, etc. This not only causes a burden of implementing
+these alerts separately in every repo, but also demands alignment between every different repo. Developers of every
+repository can differ on their opinions, leading to a huge effort of aligning between them all. Eventually, this leads
+to a situation in which it's very difficult to promote changes and new ideas that require a great amount of discussions
+and alignment.
+
+Once we have something like `kubevirt/monitoring` with a completely decoupled and autonomous from the operator, the
+need for alignment would be reduced significantly. There will still be a need to agree upon which data the operator
+exposes and how, but the monitoring logic itself will be decided in one single place.
```

Comment:
@iholder-redhat  Will this fit also small operators? Won't ir cause to much of an overhead?

Reply:
Yes.
I think the opposite is correct. The fact that the monitoring code resides within the operator's code enforces the monitoring developers to dive deep into the operator's code, go through tough code reviews, convince the maintainers of their design, etc. This enormous overhead can be minimized by simply learning the objects and their fields, developing quick with fast and simply code reviews, without a need to convince the repo maintainers / reviewers.

As this proposal points out, this is just one of the reasons to work that way, alongside keep private data safe, having better design and simpler code at the operator level, etc.